src/analysis/vector/geometry_checker/qgsgeometrycheck.cpp

```diff
@@ -175,3 +187,26 @@ double QgsGeometryCheck::scaleFactor( const QPointer<QgsVectorLayer> &layer ) co
   }
   return scaleFactor;
 }
+
+QgsGeometryCheck::Result QgsGeometryCheck::checkUniqueId( const QgsGeometryCheckerUtils::LayerFeature layerFeature, QMap< QString, QSet<QVariant> > &uniqueIds ) const
+{
+  const int uniqueIdFieldIndex = mContext->uniqueIdFieldIndex;
+  if ( uniqueIdFieldIndex == -1 )
+  {
+    return QgsGeometryCheck::Result::Success;
+  }
+
+  if ( !uniqueIds.contains( layerFeature.layerId() ) )
+  {
+    uniqueIds.insert( layerFeature.layerId(), QSet<QVariant>() );
+  }
+
+  const QVariant uniqueIdValue = layerFeature.feature().attribute( uniqueIdFieldIndex );
+  if ( uniqueIds[layerFeature.layerId()].contains( uniqueIdValue ) )
+  {
+    return QgsGeometryCheck::Result::DuplicatedUniqueId;
+  }
+  uniqueIds[layerFeature.layerId()].insert( uniqueIdValue );
```

Comment:
Let's avoid some duplicate lookups:

```suggestion
  auto uniqueIdIt = uniqueIds.find( layerFeature.layerId() );
  if ( uniqueIdIt == uniqueIds.end() )
  {
    uniqueIdIt = uniqueIds.insert( layerFeature.layerId(), QSet<QVariant>() );
  }

  const QVariant uniqueIdValue = layerFeature.feature().attribute( uniqueIdFieldIndex );
  if ( uniqueIdIt.value().contains( uniqueIdValue ) )
  {
    return QgsGeometryCheck::Result::DuplicatedUniqueId;
  }
  uniqueIdIt.value().insert( uniqueIdValue );
```